src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +32,78 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
+        {
+            if (this.BicepCompilation is null)
+            {
+                return Enumerable.Empty<IDiagnostic>();
+            }
+
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+
+            var parameters = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var parameterAssignments = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>().Where(x => this.ParamBinder.GetSymbolInfo(x) is not null);
+
+            // get diagnostics relating to missing parameter assignments or declarations
+            GetParameterMismatchDiagnostics(diagnosticWriter, parameters, parameterAssignments);
+            // get diagnostics relating to type mismatch of params between Bicep and params files
+            GetTypeMismatchDiagnostics(diagnosticWriter, parameterAssignments);
+
+            return diagnosticWriter.GetDiagnostics();
+        }
+        
+        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, ImmutableArray<Metadata.ParameterMetadata> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
+        {
+            var parametersDict = parameters.ToDictionary(x => x.Name, LanguageConstants.IdentifierComparer);
+            var parametersAssignmentsDict = parameterAssignments.ToDictionary(x => x.Name.IdentifierName, LanguageConstants.IdentifierComparer);
+            
+            // parameters that are declared but not assigned
+            var missingRequiredParams = parametersDict.Values
+                .Where(x => x.IsRequired)
+                .Where(x => !parametersAssignmentsDict.ContainsKey(x.Name))
+                .OrderBy(x => x.Name);
+            // parameters that are assigned but not declared
+            var missingAssignedParams = parametersAssignmentsDict
+                .Where(x => !parametersDict.ContainsKey(x.Key))
+                .Select(x => x.Value)
+                .OrderBy(x => x.Name.IdentifierName);
+            
+            foreach (var requiredParam in missingRequiredParams)
+            {
+                diagnosticWriter.Write(new TextSpan(0, 0), x => x.MissingParameterAssignment(requiredParam.Name));
+            }
+
+            foreach (var assignedParam in missingAssignedParams)
+            {
+                diagnosticWriter.Write(assignedParam.Span, x => x.MissingParameterDeclaration(this.ParamBinder.GetSymbolInfo(assignedParam)?.Name));
+            }
+        }
+        
+        private void GetTypeMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
+        {
+            foreach (var syntax in parameterAssignments)
+            {
+                if ((ParamsTypeManager.GetTypeInfo(syntax) is not ErrorType) && (ParamsTypeManager.GetDeclaredType(syntax) is { } declaredType) &&
+                    (!TypeValidator.AreTypesAssignable(ParamsTypeManager.GetTypeInfo(syntax), declaredType)))
```

Comment:
[nit]: The parentheses around the conditions are not necessary. Also, it would be better to put each condition on a separate line.